http4s-blaze-server/src/main/scala/com/avast/server/toolkit/http4s/middleware/CorrelationIdMiddleware.scala

```diff
@@ -0,0 +1,67 @@
+package com.avast.server.toolkit.http4s.middleware
+
+import java.util.UUID
+
+import cats.data.{Kleisli, OptionT}
+import cats.effect.Sync
+import cats.syntax.functor._
+import com.avast.server.toolkit.http4s.middleware.CorrelationIdMiddleware.CorrelationId
+import io.chrisdavenport.vault.Key
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.{Header, HttpRoutes, Request, Response}
+import org.slf4j.LoggerFactory
+
+import scala.language.higherKinds
+
+/** Provides correlation ID functionality. Either generates new correlation ID for a request or takes the one sent in HTTP header
+  * and puts it to [[org.http4s.Request]] attributes. It is also filled into HTTP response header.
+  *
+  * Use method `retrieveCorrelationId` to get the value from request attributes.
+  */
+class CorrelationIdMiddleware[F[_]: Sync](correlationIdHeaderName: CaseInsensitiveString,
```

Comment:
this class should implement some trait, so that in production code, we can program against that interface, and not this particular implementation
the good old, time proven part of dependency injection ;)

Reply:
As discussed personally I do not see the need for that, middleware is a very specific thing.